source/ShipInfoPanel.cpp

```diff
@@ -299,13 +326,88 @@ void ShipInfoPanel::UpdateInfo()
 
 void ShipInfoPanel::ClearZones()
 {
-	zones.clear();
+	zonesRight.clear();
+	zonesLeft.clear();
 	commodityZones.clear();
 	plunderZones.clear();
 }
 
 
 
+void ShipInfoPanel::SetUpHardpointCalcs(const Rectangle &bounds)
+{
+	// Figure out the left- and right-most hardpoints on the ship. If they are
+	// too far apart, the scale may need to be reduced.
+	// Also figure out how many weapons of each type are on each side.
+	maxX = 0.;
+	int count[2][2] = {{0, 0}, {0, 0}};
+	for(const Hardpoint &hardpoint : (**shipIt).Weapons())
+	{
+		// Multiply hardpoint X by 2 to convert to sprite pixels.
+		maxX = max(maxX, fabs(2. * hardpoint.GetPoint().X()));
+		++count[hardpoint.GetPoint().X() >= 0.][hardpoint.IsTurret()];
+	}
+
+	// Figure out how tall each part of the weapon listing will be.
+	int gunRows = max(count[0][0], count[1][0]);
+	int turretRows = max(count[0][1], count[1][1]);
+	// If there are both guns and turrets, add a gap of ten pixels.
+	double height = 20. * (gunRows + turretRows) + 10. * (gunRows && turretRows);
+	bool overflowsPage = height > bounds.Height();
+	height = overflowsPage ? bounds.Height() - HARDPOINT_PAGE_BUTTON_SPACE : height;
+
+	// First calculate how many pages are needed. If pages are not needed set to 1.
+	pages = overflowsPage ? 0 : 1;
+	int gunIndex = 0;
+	int turretIndex = 0;
+	rowsPerPage = static_cast<int>(height) / 10. - 1;
+	int offset = 0;
+	while(overflowsPage)
+	{
+		// First go through the guns.
+		if(gunIndex < gunRows)
+		{
+			// The guns still overflow this page.
+			if((gunRows - gunIndex) * 20. >= height)
+				gunIndex += (rowsPerPage / 2);
+			// There is space left after all guns were placed.
+			else
+			{
+				offset = gunRows - gunIndex;
+				gunIndex = gunRows;
+			}
+		}
+		// Then go through the turrets.
+		if(gunIndex >= gunRows && turretIndex < turretRows)
+		{
+			// If guns were already placed on this page.
+			if(offset)
+				// The turrets still exceed this pages space.
+				if((turretRows - turretIndex + offset) * 20. + 10. >= height)
+					turretIndex += (rowsPerPage / 2) - (offset + 1);
+				// The turrets fit this page.
+				else
+					turretIndex = turretRows;
+			else
+			{
+				// The turrets still exceed this pages space.
+				if((turretRows - turretIndex) * 20. >= height)
+					turretIndex += (rowsPerPage / 2);
+				// The turrets fit this page.
+				else
+					overflowsPage = false;
+			}
+		}
+		// Check if we went through all hardpoints.
+		if(turretIndex >= turretRows && gunIndex >= gunRows)
+			overflowsPage = false;
+		pages++;
+		offset = 0;
+	}
```

Comment:
This doesn't seem like a very good way of going about determining how many pages will be needed.

To begin with, I suggest not using a while loop.
```c++
	if(!overflowsPage)
		return;
```
Because there's nothing else to do in this method if everything fits on one page.

Next, figure out how many full pages of guns we have, and how many remaining guns we'll have afterwards:
```c++
	int gunPages = gunRows / rowsPerPage;
	int remainingGunRows = gunRos % rowsPerPage;
```
Now, let's find out if the turrets will fit on the same page as the remaining guns:
```c++
	bool turretsSharePage = (remainingGunRows + turretRows) < rowsPerPage; // We lose a row for the separation between guns and turrets, so use an exclusive comparison. This will need some slight tweaking in the final product but good enough for this example.
```
If they do fit on the same page, the total number of pages is `gunPages + 1`.
Else, we need to do the same check as we did for guns but for turrets:
```c++
	int turretPages = turretRows / rowsPerPage;
	int remainingTurretRows = turretRows % rowsPerPage;
```
And, then, the total number of pages is going to be:
```c++
	pages = gunPages + turret pages + static_cast<bool>(remainingGunRows) + static_cast<bool>(remainingTurretRows);
```

(I don't know if the code snippets here will actually work. Think of them like pseudocode.)

Reply:
I did something like that, please have a look at it